### PR TITLE
Optional bank statements

### PR DESCRIPTION
**Pull request by poma-odoo**
Bank statements are now optional in accounting. The users add statements directly and can assign statements for control reasons.
The users can enter end date and end balance real for the statements and attach the bank statement scan.

The system checks if there is a gap (lines without statement) before the current statement and warns the user
with decoration, also we check if the ending balance match the accumulated balance of the last line.

Bank statement lines now show an accumulated balance and can be sorted by drag and drop in the same date, if the user drags a line
to a position with lines of another date, it simply doesn't have any effect.
Rename accumulated balance to cumulative balance, a better name

Related enterprise PR: odoo/enterprise#30824
Task [#2879904](https://www.odoo.com/web#id=2879904&cids=1&menu_id=4720&action=4043&model=project.task&view_type=form)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_bank_statement.py [hunk 1]
```diff
@@ -815,10 +570,20 @@ def create(self, vals_list):
 
             # Otherwise field narration will be recomputed silently (at next flush) when writing on partner_id
             self.env.remove_to_compute(st_line.move_id._fields['narration'], st_line.move_id)
+
+        # No need for the user to manage their status (from Draft to Posted)
+        st_lines.move_id.action_post()
+        for st_line in st_lines.filtered('statement_id.attachment_ids'):
+            st_line._post_attachments_to_move()
```
Comment (smetl) [inline addons/account/models/account_bank_statement.py:577]: to check with PO

File: addons/account/models/account_bank_statement.py [hunk 2]
```diff
@@ -815,10 +570,20 @@ def create(self, vals_list):
...
         return st_lines
 
     def write(self, vals):
         # OVERRIDE
+        if 'date' in vals:
+            # prevents error when trying to change the date of a bank statement line and the sequence of the related
+            # move does not match the date (the date is in a different period)
+            self.move_id.name = '/'
```
Comment (smetl) [inline addons/account/models/account_bank_statement.py:585]: This diff is scarying me... :scream:

File: addons/account/views/account_bank_statement_views.xml [hunk 3]
```diff
@@ -416,12 +111,19 @@
...
         </record>
 
+        <record id="ir_cron_garbage_clean_statements" model="ir.cron">
+            <field name="name">Remove empty old bank statements</field>
+            <field name="model_id" ref="model_account_bank_statement"/>
+            <field name="state">code</field>
+            <field name="code">model._garbage_clean_statements()</field>
+            <field name="interval_type">weeks</field>
+            <field name="numbercall">-1</field>
+        </record>
```
Comment (qdp-odoo) [inline addons/account/views/account_bank_statement_views.xml:126]: to be removed. I say: we just don't care about empty statements left over.

File: addons/account/models/account_bank_statement.py (hunk 2 above)
Comment (qdp-odoo) [reply addons/account/models/account_bank_statement.py:583]: as discussed: this shouldn't be handled in this task, or specifically on the bank statement lines.

https://www.odoo.com/web#id=2976499&cids=1&menu_id=4720&action=333&active_id=967&model=project.task&view_type=form

**Review by qdp-odoo — CHANGES_REQUESTED** [review]

File: addons/account/models/account_bank_statement.py [hunk 4]
```diff
@@ -718,10 +475,92 @@ def _prepare_move_line_default_vals(self, counterpart_account_id=None):
...
+                    else:
+                        running_balance += st_line.amount
+                else:
+                    if prev_line and prev_line.statement_id:
+                        running_balance = prev_line.statement_id.balance_end_real + st_line.amount
+                    else:
+                        running_balance += st_line.amount
+                if st_line in self:
+                    st_line.running_balance = running_balance
+                prev_line = st_line
```
Comment (qdp-odoo) [inline addons/account/models/account_bank_statement.py:554]: need more comments on what you're doing

File: addons/account/models/account_bank_statement.py [hunk 5]
```diff
@@ -778,27 +639,37 @@ def _check_amounts_currencies(self):
...
-            if statement.state != 'open' and self._context.get('check_move_validity', True):
-                raise UserError(_("You can only create statement line in open bank statements."))
+            if 'statement_id' in vals:
+                statement = self.env['account.bank.statement'].browse(vals['statement_id'])
+
+                if vals.get('journal_id') and statement.journal_id and statement.journal_id.id != vals['journal_id']:
+                    # most common case, create st_line by code with default journal is not in the context,
+                    # default statement does not take journal into account because it does not have access to
+                    # the vals, it is set in the default_get and on_change is not called.
+                    del vals['statement_id']
```
Comment (qdp-odoo) [inline addons/account/models/account_bank_statement.py:661]: simplify the code and always take the journal from the statement

File: addons/account/models/account_bank_statement.py [hunk 6]
```diff
@@ -815,10 +686,18 @@ def create(self, vals_list):
...
+        # No need for the user to manage their status (from Draft to Posted)
+        st_lines.move_id.action_post()
         return st_lines
 
     def write(self, vals):
         # OVERRIDE
+        if 'date' in vals:
+            # prevents error when trying to change the date of a bank statement line and the sequence of the related
+            # move does not match the date (the date is in a different period)
+            self.move_id.name = '/'
```
Comment (qdp-odoo) [inline addons/account/models/account_bank_statement.py:699]:
```suggestion
```

**Review by qdp-odoo — CHANGES_REQUESTED** [review]
wip

**Review by qdp-odoo — CHANGES_REQUESTED** [review]
wip

File: addons/point_of_sale/models/pos_session.py [hunk 7]
```diff
@@ -183,12 +175,13 @@ def _check_start_date(self):
...
                 raise ValidationError(_("You cannot create a session before the accounting lock date."))
 
-    def _check_bank_statement_state(self):
+    def _check_bank_statement_complete(self):
         for session in self:
-            closed_statement_ids = session.statement_ids.filtered(lambda x: x.state != "open")
-            if closed_statement_ids:
-                raise UserError(_("Some Cash Registers are already posted. Please reset them to new in order to close the session.\n"
-                                  "Cash Registers: %r", list(statement.name for statement in closed_statement_ids)))
+            posted_stmt_lines = session.statement_line_ids.filtered(lambda x: x.state == "posted")
```
Comment (qdp-odoo) [inline addons/point_of_sale/models/pos_session.py:180]: statement lines don't have state anymore

**Review by qdp-odoo — CHANGES_REQUESTED** [review]
point_of_sale must be tested thoroughly: I found only one bug, and that's dubious :D

File: addons/point_of_sale/models/pos_session.py [hunk 8]
```diff
@@ -259,24 +237,22 @@ def login(self):
...
                 values['start_at'] = fields.Datetime.now()
             if session.config_id.cash_control and not session.rescue:
                 last_session = self.search([('config_id', '=', session.config_id.id), ('id', '!=', session.id)], limit=1)
-                session.cash_register_id.balance_start = last_session.cash_register_id.balance_end_real if last_session else 0
-                values['state'] = 'opening_control'
+                session.cash_register_balance_start = last_session.cash_register_balance_end_real  # defaults to 0 if lastsession is empty
             else:
                 values['state'] = 'opened'
-            session.write(values)
+            if values:
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_session.py:250]: Why did you need to change that?

File: addons/point_of_sale/models/pos_config.py [hunk 9]
```diff
@@ -219,15 +219,12 @@ def _compute_last_session(self):
         for pos_config in self:
             session = PosSession.search_read(
                 [('config_id', '=', pos_config.id), ('state', '=', 'closed')],
-                ['cash_register_balance_end_real', 'stop_at', 'cash_register_id'],
+                ['cash_register_balance_end_real', 'stop_at', 'cash_journal_id'],
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_config.py:222]: Why do you need 'cash_journal_id' ?

File: addons/point_of_sale/models/pos_session.py [hunk 10]
```diff
@@ -562,9 +551,10 @@ def get_closing_control_data(self):
             'opening_notes': self.opening_notes,
             'default_cash_details': {
                 'name': default_cash_payment_method_id.name,
-                'amount': self.cash_register_id.balance_start + total_default_cash_payment_amount +
-                                             sum(self.cash_register_id.line_ids.mapped('amount')),
-                'opening': self.cash_register_id.balance_start,
+                'amount': self.cash_register_balance_start
+                          + total_default_cash_payment_amount
+                          + self.cash_real_transaction,
```
Comment (caburj) [inline addons/point_of_sale/models/pos_session.py:556]:
```suggestion
                          + sum(self.statement_line_ids.mapped('amount')),
```

File: addons/account/models/account_bank_statement.py [hunk 11]
```diff
@@ -1,476 +1,167 @@
...
+            defaults['balance_start'] = self.env['account.bank.statement.line'].search(
+                domain=[
+                    ('date', '<=', self._context['st_line_date']),
+                    ('journal_id', '=', self._context.get('st_line_journal_id')),
+                ],
+                order='internal_index desc',
+                limit=1
+            ).running_balance
+        # creating statements with split button
+        elif self._context.get('split_line_id'):
```
[the remark this answers is not in the document]
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py]: Then we will need a mode and different data, this way we have fewer items in the context.

File: addons/account/models/account_bank_statement.py (hunk 1 above)
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py]: fixed

File: addons/account/tests/test_account_bank_statement.py [hunk 12]
```diff
@@ -847,43 +526,147 @@ def test_zero_amount_statement_line(self):
...
+        self.env.user.company_id = self.company_data_2['company']
+
+        line7 = self.create_bank_transaction(7, '2020-01-10', sequence=1)
+        line8 = self.create_bank_transaction(8, '2020-01-10', sequence=2)
+        line2 = self.create_bank_transaction(2, '2020-01-13')
+        _line6 = self.create_bank_transaction(6, '2020-01-11')
+        line5 = self.create_bank_transaction(5, '2020-01-12', sequence=3)
+        line4 = self.create_bank_transaction(4, '2020-01-12', sequence=2)
+        _line1 = self.create_bank_transaction(1, '2020-01-13')
+        line3 = self.create_bank_transaction(3, '2020-01-12', sequence=1)
```
[the remark this answers is not in the document]
Comment (poma-odoo) [reply addons/account/tests/test_account_bank_statement.py:723]: the names indicate the original order of the lines

File: addons/account/models/account_bank_statement.py (hunk 2 above)
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py:671]: Is there any alternative, (other than erasing a move and recreate it, make a move draft and repost it)? The last discussion resulted in the user can renumber the moves, so having gaps is not a problem. And both this way and setting the move to draft and repost it or recreating it does the same.
So I leave it like this until the forementioned task sees its end.

File: addons/account/models/account_bank_statement.py [hunk 13]
```diff
@@ -1,476 +1,245 @@
...
+                domain=[
+                    ('internal_index', '<=', current_st_line.internal_index),
+                    ('internal_index', '>', line_before.internal_index or ''),
+                    ('journal_id', '=', current_st_line.journal_id.id),
+                ],
+                order='internal_index desc',
+            )
+        # if it is called from the action menu, we can have both start and end balances and we filter out
+        # completed statements, because it is probably due to a mistake from the user
+        elif self._context.get('active_model') == 'account.bank.statement.line' and self._context.get('active_ids'):
```
[the remark this answers is not in the document]
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py:181]: `'account.bank.statement.line'` and `'active_ids'` are filled automatically in kanban, so in both cases they are available.
I could use something like `is_split` instead of `split_line_id` and use active_id, but I think this way is more explicit and makes debugging easier.

File: addons/account/models/account_bank_statement.py [hunk 14]
```diff
@@ -1,476 +1,245 @@
...
-            statement.total_entry_encoding = sum([line.amount for line in statement.line_ids])
-            statement.balance_end = statement.balance_start + statement.total_entry_encoding
-            statement.difference = statement.balance_end_real - statement.balance_end
+    # Used to hold the reference of the external mean that created this statement (name of imported file,
+    # reference of online synchronization...)
+    reference = fields.Char(
+        string='External Reference',
+        copy=False,
+    )
+    date = fields.Date(compute='_compute_date_index', store=True,)
```
[the remark this answers is not in the document]
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py]: SO it is automatically filled, cleaner interface, fewer things to worry/make mistake for the user. Also, the default value won't be known because usually the users start from the earliest line to encode a statement, so if we set a default value based on the first line, it would be wrong most of the time.

File: addons/point_of_sale/models/pos_session.py [hunk 15]
```diff
@@ -1397,12 +1397,13 @@ def _amount_converter(self, amount, date, round):
         return self.currency_id._convert(amount, self.company_id.currency_id, self.company_id, date, round=round)
 
     def show_cash_register(self):
```
Comment (caburj) [inline addons/point_of_sale/models/pos_session.py:1399]: This method is now unused. But since you properly adapted the returned value, perhaps you can restore the button you removed from [pos_session_view.xml](https://github.com/odoo/odoo/pull/99092/files#diff-97e241e39c63c662bf535d48566bc0c8f26724826b79faec4af6c9b67148cb9e)?

File: addons/point_of_sale/models/pos_session.py (hunk 10 above)
Comment (caburj) [reply addons/point_of_sale/models/pos_session.py:555]: Note the `cash_real_transaction` is not yet assigned at this point. The enclosing method is called prior to calling `_validate_session`.

File: addons/account/models/account_bank_statement.py (hunk 5 above)
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py:633]: Are you sure about this? If the statement is not comply with the journal things would be weired, for example if the user saves the default journal in `ir.default`

File: addons/account/models/account_bank_statement.py (hunk 6 above)
Comment (poma-odoo) [reply addons/account/models/account_bank_statement.py:671]: Considering that the sequence is converted to yearly, It will only disable editing of the lines date between years, the user can still erase the line and recreate it, so nothing critical, I am removing this.

File: addons/point_of_sale/models/pos_session.py (hunk 8 above)
Comment (poma-odoo) [reply addons/point_of_sale/models/pos_session.py]: I cannot find this diff

File: addons/point_of_sale/models/pos_config.py (hunk 9 above)
Comment (poma-odoo) [reply addons/point_of_sale/models/pos_config.py:222]: Nope, it is the result of replace all, it looks like `cash_register_id` was not used here

File: addons/point_of_sale/models/pos_session.py (hunk 15 above)
Comment (poma-odoo) [reply addons/point_of_sale/models/pos_session.py:1399]: The problem is we don't have this view (tree view of transactions) in the community version, it is now an enterprise only. If you think this button is necessary, I can add it in the enterprise, if you think it is not that important I can erase this method for good

Comment (caburj) [reply addons/point_of_sale/models/pos_session.py:1399]: Would be nice to keep it in enterprise so that it will look like we didn't remove a feature. Would you agree @pimodoo? Or can we just let the button go?

Comment (pimodoo) [reply addons/point_of_sale/models/pos_session.py:1384]: As it exists, it could be put in enterprise. By saying it only exists in enterprise, do you mean it is in the module account_accountant? because we don't depend on it. So if it means we should do create a module pos_account_accountant, we probably should not keep it.

Comment (poma-odoo) [reply addons/point_of_sale/models/pos_session.py:1384]: Yes it is in account_accountant
